File: 873/length-of-longest-fibonacci-subsequence.js
/**
 * @param {number[]} A
 * @return {number}
 */
var lenLongestFibSubseq = function (A) {
    let len = A.length;
    let hash = {};
    for (let i = 0; i < len; i++) {
        hash[A[i]] = i;
    }
    let dp = [...Array(len)].map(r => Array(len).fill(2));
    let ans = 0;
    for (let j = 0; j < len; j++) {
        for (let k = j+1; k < len; k++) {
            let a_i = A[k] - A[j];
            if (a_i >= A[j]) break;
            if (hash[a_i] == undefined) continue;
            let i = hash[a_i];
            dp[j][k] = dp[i][j] + 1;
            ans = ans > dp[j][k] ? ans : dp[j][k];
        }
    }
    return ans;
};

function lenLongestFibSubseq(A) {
    let len = A.length;
    let s = new Set();
    for (let item of A) {
        s.add(item);
    } 
    let ans = 0;
    for (let i = 0; i < len; i++) {
        for (let j = i+1; j < len; j++) {
            let [a, b, c] = [A[i], A[j], a+b];
            let c = a + b;
            let l = 2;
            while (s.has(c)) {
                a = b;
                b = c;
                c = a+b;
                l++;
                ans = l > ans ? l : ans;
            }
        }
    }
    return ans;
}

File: 873/length-of-longest-fibonacci-subsequence.go
package main

import "fmt"

func lenLongestFibSubseq(A []int) int {
	size := len(A)
	var hash map[int]int
	hash = make(map[int]int)
	for i, item := range A {
		hash[item] = i
	}

	// dp := make([][]int, size)
	// for i := range dp {
	// 	dp[i] = make([]int, size)
	// }
	// dp := [size][size]int{}
	dp := make([][]int, size)
	for i := 0; i < size; i++ {
		row := make([]int, size)
		for j := 0; j < size; j++ {
			row[j] = 2
		}
		dp[i] = row
		// dp = append(dp, row)
	}

	ans := 0
	for j := 0; j < size; j++ {
		for k := j + 1; k < size; k++ {
			a_i := A[k] - A[j]
			if a_i >= A[j] {
				break
			}
			i, ok := hash[a_i]
			if !ok {
				continue
			}
			dp[j][k] = dp[i][j] + 1
			if dp[j][k] > ans {
				ans = dp[j][k]
			}
		}
	}
	return ans
}

func main() {
	A := []int{1, 2, 3, 4, 5}
	fmt.Println(lenLongestFibSubseq(A))
}
